Set ESlint rules more strict 🚑

ndelangen opened the pull request:
Issue: eslint rules are too loose, errors are passing though.

## What I did
As described in review comments #904 linting is rather non-strict right now.
Though I'm personally of the opinion too strict linting can be harmful, I agree, we could make it a bit more strict to weed out some bad code we have and prevent bad come from getting pushed to master.

- [x] use eslint to lint all `.js` & `.jsx` files
- [x] use eslint to lint all `.json` files
- [x] sort all package.json using `sort-package-json` (single time).
- [x] set rules to a much much much stricter set
- [x] fix all js-files errors
- [x] fix all json-files errors 
- [x] use remark-lint to lint all `.md` files
- [x] use remark-lint-code-eslint to lint all code-samples inside markdown files
- [x] use prettier
- [x] fix all markdown-files errors
- [x] setup `lint-staged`
- [x] test `lint-staged`
- [x] fix any regressions
- [x] more covfefe

## How to test
run `npm run lint`

ndelangen committed:
ADD stricter eslint rules

ndelangen committed:
FIX linting

ndelangen committed:
FIX for https://github.com/storybooks/storybook/commit/fc7e09b810678c8db2235c397a49df35dc667ef2?diff=split#commitcomment-21940155

subpopular commented on packages/addon-info/src/components/PropVal.js:
```diff
@@ -107,7 +107,7 @@ function previewProp(val) {
 
   if (!braceWrap) return content;
-  return <span>{{ content }}</span>;
+  return <span>{`{ ${content} }`}</span>;
```
nice!

ndelangen committed:
Merge branch 'master' into eslint-strict

# Conflicts:
#	packages/addon-graphql/package.json
#	packages/addon-info/package.json
#	packages/addon-info/src/components/Node.js
#	packages/addon-knobs/package.json
#	packages/addon-notes/package.json
#	packages/channel-postmessage/package.json
#	packages/channel-websocket/package.json
#	packages/decorator-centered/package.json
#	packages/react-native-storybook/package.json
#	packages/react-storybook/src/server/config.js
#	packages/storybook-ui/package.json
#	packages/storybook-ui/src/modules/ui/components/layout/index.js

ndelangen committed:
FIX linting

theinterned committed:
fixes linting in storyshots

theinterned commented:
question: I've got this error:

```console
/Users/user/src/theinterned-storybook/packages/storybook-ui/src/modules/ui/components/left_panel/text_filter.test.js
  1:1  error  'enzyme' should be listed in the project's dependencies. Run 'npm i -S enzyme' to add it  import/no-extraneous-dependencies
```
this is due to enzyme being in the root `package.json` is it appropriate just add an `eslint-disable-line` or should we modify the eslint-config?

theinterned committed:
FIX: listing in the `storybook-ui` package

notes:

1. did not resolve `jsx-a11y` errors as that would require closer
attention to the browser
2. did not resolve any `'enzyme' should be listed in the project's
dependencies. Run 'npm i -S enzyme' to add it
import/no-extraneous-dependencies` see note:
https://github.com/storybooks/storybook/pull/911#issuecomment-298077570

theinterned committed:
fix: failing Panel.js test in add-knobs

adds `global` as the  dep was missing in add-on-knobs

theinterned committed:
resolves listing in Stories component of storybook-ui

didn’t resolve jsx-a11y linting in this pass

theinterned committed:
fix: test for storybook-ui let_panel/index

uses `toMatchObject` rather than `toEqual` to be more lenient about
what props are present

theinterned committed:
fix: eslint -- resolves globals using the `global` module rather than the eslint config for globals

theinterned committed:
fix: linting in `react-storybook` package

theinterned committed:
fix: adds missing `ws` dep to `react-native-storybook`

theinterned committed:
fix: adds missing `global` dep to `react-native-storybook`

theinterned committed:
fix: linting in `react-native-storybook` package

theinterned committed:
fix: listing in `getstorybook` package

theinterned committed:
fix: adds `react: "*"` as a peer dep for `decorator-cantered` to fix `import/no-extraneous-dependencies` linter warning

theinterned committed:
fix: adds `global` dep to `channel-postmessage` and `channel-websocket` to resolve `import/no-extraneous-dependencies` lint issue

theinterned committed:
fix: linting in `channel-websockets`

theinterned committed:
fix: peer dep for `@kadira/storybook-addons` wasn't resolving properly by eslint

theinterned committed:
fix: listing of `add-on-notes`

theinterned committed:
fix: linting for `addon-knobs` package

note: didn’t tackle jsx-a11y on this pass.

theinterned commented:
@UsulPro has fixed some issues in the **addon-info** package in this PR: https://github.com/storybooks/storybook/pull/980

theinterned committed:
fix: listing for `addon-info` package

theinterned committed:
fix: adds `global` dep to resolve linting

theinterned committed:
fix: linting `no-underscore-dangle` faux-private members in `addons` package

theinterned committed:
fix: linting in `addon-graphql` package

theinterned committed:
fix: linting in `addon-comments` package

theinterned commented:
I thought I'd give an update on my work. 

I'm down to `✖ 72 problems (64 errors, 8 warnings)`

For the time being I'm skipping jsx-a11y issues as I didn't feel comfortable fixing a number of the UI-related issues without running the "kitchen-sink" project and generally making more nuanced decisions. I plan to tackle that in a separate PR.

My PR is here https://github.com/storybooks/storybook/pull/1000

@ndelangen does it make sense to close that one in favour of this?

ndelangen commented:
closed in favour of https://github.com/storybooks/storybook/pull/1000

ndelangen committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
CHANGED eslint-config-airbnb for a fork waiting for publish
of https://github.com/airbnb/javascript/commit/121a95d71524032bfa858271b318aa6e19553dfc

UPDATED dependencies
CHANGED devDependencies linting
ADD enzyme and storybook-addon to devDependencies in a few packages

ndelangen committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
Merge branch 'master' into eslint-strict

# Conflicts:
#	addons/centered/package.json
#	addons/graphql/package.json
#	addons/info/package.json
#	addons/knobs/package.json
#	addons/links/.storybook/config.js
#	addons/notes/package.json
#	addons/notes/src/index.js
#	addons/notes/src/register.js
#	addons/options/src/manager/index.js
#	addons/storyshots/src/index.js
#	addons/storyshots/stories/required_with_context/Button.stories.js
#	addons/storyshots/stories/required_with_context/Welcome.js
#	app/react-native/package.json
#	app/react-native/src/manager/index.js
#	lib/channel-postmessage/package.json
#	lib/channel-postmessage/src/index.js
#	lib/channel-websocket/package.json
#	lib/channel-websocket/src/index.js
#	lib/cli/generators/METEOR/index.js
#	lib/cli/generators/REACT/template/stories/Welcome.js
#	lib/cli/generators/REACT_SCRIPTS/index.js
#	lib/cli/generators/WEBPACK_REACT/template/stories/Welcome.js
#	packages/react-storybook/src/server/track_usage.js
#	packages/storybook-ui/src/modules/ui/components/layout/usplit.js
#	scripts/prepublish.js

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
Merge branch 'master' into eslint-strict

# Conflicts:
#	.eslintignore
#	.travis.yml
#	addons/actions/src/containers/ActionLogger/index.js
#	addons/centered/package.json
#	addons/graphql/package.json
#	addons/info/src/components/PropTable.js
#	addons/info/src/components/Props.js
#	addons/knobs/package.json
#	addons/knobs/src/components/Panel.js
#	addons/notes/package.json
#	addons/notes/src/index.js
#	addons/notes/src/register.js
#	addons/storyshots/package.json
#	addons/storyshots/src/index.js
#	app/react-native/package.json
#	app/react/package.json
#	examples/cra-storybook/package.json
#	examples/cra-storybook/src/stories/Welcome.js
#	examples/test-cra/src/stories/Button.js
#	examples/test-cra/src/stories/Welcome.js
#	lib/channel-postmessage/package.json
#	lib/channel-websocket/package.json
#	lib/cli/bin/generate.js
#	lib/cli/generators/METEOR/index.js
#	lib/cli/generators/REACT/index.js
#	lib/cli/generators/REACT/template/stories/index.js
#	lib/cli/generators/REACT_NATIVE/index.js
#	lib/cli/generators/REACT_SCRIPTS/index.js
#	lib/cli/generators/REACT_SCRIPTS/template/src/stories/Welcome.js
#	lib/cli/generators/WEBPACK_REACT/index.js
#	lib/cli/lib/helpers.js
#	lib/cli/package.json
#	package.json

ndelangen committed:
Mostly re-applied changes

ndelangen committed:
FIX linting

ndelangen commented:
Re-opening, will be closing #1000

ndelangen committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
ADD lint-staged

shilman committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
Set npm needed for project to lower then 5.0.0

ndelangen committed:
FIX unit tests snapshots && FIX linting

shilman reviewed (approved):
@ndelangen this is looking much better. there is still some weirdness with the `prettier` config, but I couldn't figure out how to get rid of it:

```
/Users/shilman/projects/storybook/new/storybook/lib/ui/src/modules/ui/components/shortcuts_help.js
   69:53  warning  Delete `·(`  prettier/prettier
   75:3   warning  Delete `)`   prettier/prettier
   ...
```

Also when I run `npm run lint:js -- --fix` it removes a bunch of these parens and makes the code harder to read.

Finally, there are some warnings when I run the test suite.

Nevertheless, I think this is good to merge. What do you think?

ndelangen committed:
FIX actions addon register
ADD dev script
RENAME addon-namespaces
CHANGE lint:markdown to lint:md

ndelangen committed:
FIX linting

ndelangen committed:
FIX eslint cache location

ndelangen committed:
Merge branch 'master' into eslint-strict

ndelangen committed:
FIX after merge

ndelangen committed:
ADD eslint config for examples inside markdown

FIX linting

ndelangen committed:
FIX changelog markdown

ndelangen committed:
FIX linting && ADD globals to makrdown code examples lint config

ndelangen committed:
FIX linting

shilman commented on .remarkrc:
```diff
@@ -5,7 +5,8 @@
       "options": {
-        "fix": true
+        "fix": true,
+        "configFile": "/Users/dev/Projects/GitHub/storybook/react-storybook/.eslintrc-markdown.js"
```
@ndelangen ???

ndelangen replied:
Will fix this

shilman commented on docs/pages/addons/api/index.md:
````diff
@@ -26,7 +26,7 @@ See how we can use this:
 ```js
 // Register the addon with a unique name.
-addonAPI.register('kadira/notes', (storybookAPI) => {
+addonAPI.register('my-organisation/my-addon', (storybookAPI) => {
````
@ndelangen can we put this in a different PR? i think it is unrelated to eslint?

shilman reviewed (changes requested):
Hey this looks great but I think we should get it merged and then do refinements on the merged version. I think this PR is getting out of control?

shilman committed:
Add missing import

+ make listing more specific

shilman reviewed (changes requested):
@ndelangen I am seeing an error trying to run gatsby in the docs

```
npm run dev
```

Huge string of errors:
```
ERROR in ./~/css-loader!./~/postcss-loader/lib!./components/Docs/Content/style.css
Module build failed: Error: No PostCSS Config found in: /Users/shilman/projects/storybook/new/storybook/docs/components/Docs/Content
    at Error (native)
    at /Users/shilman/projects/storybook/new/storybook/docs/node_modules/postcss-load-config/index.js:51:26

 @ ./components/Docs/Content/style.css 4:14-130 18:2-22:4 19:20-136
...
```

ndelangen committed:
ADD linting on some rc files, CHANGe to js rc for remark

ndelangen committed:
FIX linting

ndelangen committed:
FIX linting

ndelangen committed:
ADD eslint config for .json files && FIX linting

ndelangen committed:
Merge branch 'master' into eslint-strict

# Conflicts:
#	app/react-native/src/manager/components/PreviewHelp.js

ndelangen committed:
CHANGE all package.jsons so they are uniformly sorted

ndelangen committed:
FIX docs

ndelangen committed:
CHANGE path to markdown code example eslint config to relative path

ndelangen commented:
It's fixed @shilman !

ndelangen commented:
I'm ready to merge this, who can test locally and approve?

ndelangen committed:
Merge branch 'master' into eslint-strict
